src/java/org/apache/cassandra/utils/concurrent/ExtendedNonBlockingHashMap.java

```diff
@@ -0,0 +1,181 @@
+{
+    /**
+     * Get a value for a given key, waiting for initialization if necessary. This method does not initialize the value
+     * if missing. It returns {@code null} regardless the value is missing or failed to initialize.
+     */
+    public V getIfReady(K key)
+    {
+        Future<V> future = get(key);
+        if (future != null && future.isDone())
+            return future.getNow();
+        else
+            return null;
+    }
+
+    /**
+     * If the value for the given key is missing, execute a load function to obtain a value and put it into the map.
+     * It is guaranteed that the load function will be executed only once when the key is missing and mulitple threads
+     * called this method for the same key.
+     * <p>
+     * When the mapping function returns {@code null}, {@link NullPointerException} is thrown. When the mapping function
+     * throws exception, it is rethrown by this method. In both cases nothing gets added to the map.
+     */
+    public V blockingLoadIfAbsent(K key, Supplier<? extends V> loadFunction) throws RuntimeException
+    {
+        while (true)
+        {
+            Future<V> future = get(key);
+            AsyncPromise<V> newEntry = null;
+            if (future == null)
+            {
+                newEntry = new AsyncPromise<>();
+                future = putIfAbsent(key, newEntry);
+                if (future == null)
+                {
+                    // We managed to create an entry for the value. Now initialize it.
+                    future = newEntry;
+                    try
+                    {
+                        V v = loadFunction.get();
+                        if (v == null)
+                            throw new NullPointerException("The mapping function returned null");
+                        else
+                            newEntry.setSuccess(v);
+                    }
+                    catch (Throwable t)
+                    {
+                        newEntry.setFailure(t);
+                        // Remove future so that construction can be retried later
+                        remove(key, future);
+                    }
+                }
+                else
+                {
+                    newEntry = null;
+                }
+
+                // Else some other thread beat us to it, but we now have the reference to the future which we can wait for.
+            }
+
+            future.awaitUninterruptibly();
+
+            if (future.getNow() != null) // implies success
+                return future.getNow();
+
+                // if blockingUnloadIfPresent was called in the meantime (success and getNow == null), we simply retry hoping that unloading gets finished soon
+                // also we retry if the concurrent attempt to load entry failed - but we do not retry if this attempt failed (failed and newEntry == null)
+            else if (future.isSuccess() || newEntry == null)
```

Comment:
I think this would be clearer as
```
if (newEntry != null)
    future.rethrowIfFailed();

Thread.yield();
```
I would also consider using a boolean (named e.g. `usingOurFuture`) instead of `newEntry != null`.

Reply:
great idea!